### PR TITLE
Include "simple" UTF-8 validation and transcoding logic for interpreted and low-footprint scenarios

### PR DESCRIPTION
**Pull request by pavelsavara**
trying to run unit tests on 
https://github.com/dotnet/runtime/issues/48006

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=49372) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=49372) before we can accept your contribution.<br/><br/>:x: pavelsavara [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=49372)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=49372) it.</sub>

**Comment by GrabYourPitchforks** [comment]
Left some minor comments, but the code and the approach generally LGTM. But I'd like to see further discussion on https://github.com/dotnet/runtime/issues/48006 as to whether it's even worthwhile to take this, and if so on what platforms we should conditionally compile this in. The initial platform list proposed in the .csproj here seems quite reasonable.

File: src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems [hunk 1]
```diff
@@ -15,6 +15,7 @@
     <SupportsX86Intrinsics Condition="'$(Platform)' == 'x64' or ('$(Platform)' == 'x86' and '$(TargetsUnix)' != 'true')">true</SupportsX86Intrinsics>
     <ILLinkSharedDirectory>$(MSBuildThisFileDirectory)ILLink\</ILLinkSharedDirectory>
     <Is64Bit Condition="'$(Platform)' == 'arm64' or '$(Platform)' == 'x64'">true</Is64Bit>
+    <OptimizeForSize Condition="'$(TargetsBrowser)' == 'true' or '$(TargetsAndroid)' == 'true' or '$(TargetsiOS)' == 'true' or '$(TargetstvOS)' == 'true'">true</OptimizeForSize>
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems:18]: Both Android and iOS phones are higher-end hardware. I would expect we want to keep the vectorized implementations for those.

Note that we are enabling ARM64 intrinsics for those platforms: https://github.com/dotnet/runtime/search?q=ARM64+intrinsics+Mono&type=issues

**Comment by pavelsavara** [comment]
Hi guys, 

I'm learning dotnet contributor ropes on this task, so please bear with me. 
I'm still struggling with running it and debugging it locally, so this draft PR was just easy way for me to run unit tests on it and see what happens. So it was just blind shot yesterday and there were errors in the logic. 

I will further improve it based on the feedback, thanks for it.

Regarding question, if this should be merged at all, I thought we could just try and see if the change makes it actually smaller.
I still don't know where to look for the binaries to compare, but eventually I would get there.

There is also Utf16Utility shall we consider those too ? 
Remaining methods in Utf8Utility look small enough to me, right ?

Thanks!

File: src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems (hunk 1 above)
Comment (pavelsavara) [reply src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems]: Reduced to browser and tvOS only

File: src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems [hunk 2]
```diff
@@ -1170,6 +1168,14 @@
       <Link>Common\System\Threading\Tasks\TaskToApm.cs</Link>
     </Compile>
   </ItemGroup>
+  <ItemGroup Condition="'$(OptimizeForSize)' != true">
```
Comment (GrabYourPitchforks) [inline src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems:1171]: Nit: We put single quotes around true. Same comment applies to line 1176.
(See line 21 for example.)

File: src/libraries/System.Private.CoreLib/src/System/Text/Unicode/Utf8Utility.SizeOpt.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+            int cumulativeUtf16CodeUnitCount = 0;
+            int cumulativeBytesConsumed = 0;
+            int cumulativeScalarValueCount = 0;
+            while (!input.IsEmpty)
+            {
+                if (Rune.DecodeFromUtf8(input, out Rune rune, out int bytesConsumed) != OperationStatus.Done)
+                    break;
+                input = input.Slice(bytesConsumed);
+                cumulativeUtf16CodeUnitCount += rune.Utf16SequenceLength;
+                cumulativeBytesConsumed += bytesConsumed;
```
Comment (GrabYourPitchforks) [inline src/libraries/System.Private.CoreLib/src/System/Text/Unicode/Utf8Utility.SizeOpt.cs:99]: You don't need this counter. You can calculate bytes consumed after the loop as "original input length - remaining length of 'input' span".

**Review by GrabYourPitchforks — APPROVED** [review]
LGTM, and thanks for your contribution! I'll defer to other more knowledgeable people on whether the switched platforms are correct. (To those reviewers, check the .csproj.)

**Pull request merged**